[functorch] Add a bunch of low hanging fruit linalg batch rules

samdow opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82325
* #82814
* __->__ #82177

This gets ~9 batching rules using only decomps and existing macros so I stuck them all together, happy to break it up if there's a more logical way.

Also gets slogdet and solves for https://github.com/pytorch/functorch/issues/984. I haven't profiled locally but I will

Commit:
[functorch] Add a bunch of low hanging fruit linalg batch rules

[ghstack-poisoned]

Commit:
Update on "[functorch] Add a bunch of low hanging fruit linalg batch rules"


This gets ~9 batching rules using only decomps and existing macros so I stuck them all together, happy to break it up if there's a more logical way.

Also gets slogdet and solves for https://github.com/pytorch/functorch/issues/984. I haven't profiled locally but I will

[ghstack-poisoned]

zou3519 commented on functorch/functorch/csrc/BatchRulesLinearAlgebra.cpp:
```diff
@@ -233,6 +233,20 @@ struct LinalgCheckMatrixBinaryRuleHelper;
+      const Tensor& first, optional<int64_t> first_bdim,
+      const Tensor& second, optional<int64_t> second_bdim,
+      T... extra_args) {
+    TORCH_CHECK(!first_bdim.has_value() || first.dim() > 2,
+              op_name, ": The input tensor A must have at least 2 dimensions.");
+    TORCH_CHECK(!second_bdim.has_value() || second.dim() > 2,
+              op_name, ": The input tensor A must have at least 2 dimensions.");
+    const auto tensor_other = _binary_pointwise_helper(first, first_bdim, second, second_bdim);
```
Same comments as the last PR: should compute rank without bdim and use that, and also binary_pointwise_helper does type promotion

Commit:
Update on "[functorch] Add a bunch of low hanging fruit linalg batch rules"


This gets ~9 batching rules using only decomps and existing macros so I stuck them all together, happy to break it up if there's a more logical way.

Also gets slogdet and solves for https://github.com/pytorch/functorch/issues/984. I haven't profiled locally but I will

[ghstack-poisoned]

Commit:
Update on "[functorch] Add a bunch of low hanging fruit linalg batch rules"


This gets ~9 batching rules using only decomps and existing macros so I stuck them all together, happy to break it up if there's a more logical way.

Also gets slogdet and solves for https://github.com/pytorch/functorch/issues/984. I haven't profiled locally but I will

[ghstack-poisoned]

Commit:
Update on "[functorch] Add a bunch of low hanging fruit linalg batch rules"


This gets ~9 batching rules using only decomps and existing macros so I stuck them all together, happy to break it up if there's a more logical way.

Also gets slogdet and solves for https://github.com/pytorch/functorch/issues/984. I haven't profiled locally but I will

[ghstack-poisoned]

Commit:
Update on "[functorch] Add a bunch of low hanging fruit linalg batch rules"


This gets ~9 batching rules using only decomps and existing macros so I stuck them all together, happy to break it up if there's a more logical way.

Also gets slogdet and solves for https://github.com/pytorch/functorch/issues/984. I haven't profiled locally but I will

[ghstack-poisoned]

Commit:
Update on "[functorch] Add a bunch of low hanging fruit linalg batch rules"


This gets ~9 batching rules using only decomps and existing macros so I stuck them all together, happy to break it up if there's a more logical way.

Also gets slogdet and solves for https://github.com/pytorch/functorch/issues/984. I haven't profiled locally but I will

[ghstack-poisoned]

Commit:
Update on "[functorch] Add a bunch of low hanging fruit linalg batch rules"


This gets ~9 batching rules using only decomps and existing macros so I stuck them all together, happy to break it up if there's a more logical way.

Also gets slogdet and solves for https://github.com/pytorch/functorch/issues/984. I haven't profiled locally but I will

[ghstack-poisoned]

Commit:
Update on "[functorch] Add a bunch of low hanging fruit linalg batch rules"


This gets ~9 batching rules using only decomps and existing macros so I stuck them all together, happy to break it up if there's a more logical way.

Also gets slogdet and solves for https://github.com/pytorch/functorch/issues/984. I haven't profiled locally but I will

[ghstack-poisoned]

Commit:
Update on "[functorch] Add a bunch of low hanging fruit linalg batch rules"


This gets ~9 batching rules using only decomps and existing macros so I stuck them all together, happy to break it up if there's a more logical way.

Also gets slogdet and solves for https://github.com/pytorch/functorch/issues/984. I haven't profiled locally but I will

[ghstack-poisoned]

Commit:
Update on "[functorch] Add a bunch of low hanging fruit linalg batch rules"


This gets ~9 batching rules using only decomps and existing macros so I stuck them all together, happy to break it up if there's a more logical way.

Also gets slogdet and solves for https://github.com/pytorch/functorch/issues/984. I haven't profiled locally but I will

[ghstack-poisoned]

Commit:
Update on "[functorch] Add a bunch of low hanging fruit linalg batch rules"


This gets ~9 batching rules using only decomps and existing macros so I stuck them all together, happy to break it up if there's a more logical way.

Also gets slogdet and solves for https://github.com/pytorch/functorch/issues/984. I haven't profiled locally but I will

[ghstack-poisoned]

samdow commented:
@pytorchbot merge

Commit:
Update on "[functorch] Add a bunch of low hanging fruit linalg batch rules"


This gets ~9 batching rules using only decomps and existing macros so I stuck them all together, happy to break it up if there's a more logical way.

Also gets slogdet and solves for https://github.com/pytorch/functorch/issues/984. I haven't profiled locally but I will

[ghstack-poisoned]

samdow commented:
@pytorchbot merge
